bindings/matrix-sdk-crypto-js/src/backup_recovery_key.rs

```diff
@@ -0,0 +1,176 @@
+    #[wasm_bindgen(js_name = "fromBase64")]
+    pub fn from_base64(key: String) -> Result<BackupRecoveryKey, JsError> {
+        Ok(Self { inner: RecoveryKey::from_base64(&key)?, passphrase_info: None })
+    }
+
+    /// Try to create a [`BackupRecoveryKey`] from a base 58 encoded string.
+    #[wasm_bindgen(js_name = "fromBase58")]
+    pub fn from_base58(key: String) -> Result<BackupRecoveryKey, JsError> {
+        Ok(Self { inner: RecoveryKey::from_base58(&key)?, passphrase_info: None })
+    }
+
+    /// Create a new [`BackupRecoveryKey`] from the given passphrase.
+    #[wasm_bindgen(js_name = "newFromPassphrase")]
+    pub fn new_from_passphrase(passphrase: String) -> BackupRecoveryKey {
+        let mut rng = thread_rng();
+        let salt: String = iter::repeat(())
+            .map(|()| rng.sample(Alphanumeric))
+            .map(char::from)
+            .take(Self::SALT_SIZE)
+            .collect();
+
+        BackupRecoveryKey::from_passphrase(passphrase, salt, Self::PBKDF_ROUNDS)
+    }
+
+    /// Restore a [`BackupRecoveryKey`] from the given passphrase.
+    #[wasm_bindgen(js_name = "fromPassphrase")]
+    pub fn from_passphrase(passphrase: String, salt: String, rounds: i32) -> Self {
+        let mut key = Box::new([0u8; Self::KEY_SIZE]);
+        let rounds = rounds as u32;
+
+        pbkdf2::<Hmac<Sha512>>(passphrase.as_bytes(), salt.as_bytes(), rounds, key.deref_mut());
+
+        let recovery_key = RecoveryKey::from_bytes(&key);
+
+        key.zeroize();
+
+        Self {
+            inner: recovery_key,
+            passphrase_info: Some(PassphraseInfo {
+                private_key_salt: salt.into(),
+                private_key_iterations: rounds as i32,
+            }),
+        }
+    }
+
+    /// Convert the recovery key to a base 58 encoded string.
+    #[wasm_bindgen(js_name = "toBase58")]
+    pub fn to_base58(&self) -> JsString {
+        self.inner.to_base58().into()
+    }
+
+    /// Convert the recovery key to a base 64 encoded string.
+    #[wasm_bindgen(js_name = "toBase64")]
+    pub fn to_base64(&self) -> JsString {
+        self.inner.to_base64().into()
+    }
+
+    /// Get the public part of the backup key.
+    #[wasm_bindgen(getter, js_name = "megolmV1PublicKey")]
+    pub fn megolm_v1_public_key(&self) -> MegolmV1BackupKey {
+        let public_key = self.inner.megolm_v1_public_key();
+
+        MegolmV1BackupKey { inner: public_key, passphrase_info: self.passphrase_info.clone() }
+    }
+
+    /// Try to decrypt a message that was encrypted using the public part of the
+    /// backup key.
+    #[wasm_bindgen(js_name = "decryptV1")]
+    pub fn decrypt_v1(
+        &self,
+        ephemeral_key: String,
+        mac: String,
+        ciphertext: String,
+    ) -> Result<JsValue, JsError> {
+        self.inner
+            .decrypt_v1(&ephemeral_key, &mac, &ciphertext)
+            .map_err(|e| e.into())
+            .map(|r| JSON::parse(&r).unwrap())
```

Comment:
Another one. Hmm, this now assumes that whatever you are decrypting is JSON, I guess that's currently true but worries me slightly.

Reply:
let's just return the JSON. The javascript side can deserialise it more easily anyway.